### PR TITLE
Add snapshot storage

### PR DESCRIPTION
**Pull request by Anton-Rampage**

### CODE REVIEW

File: pkg/state/snapshot_storage.go [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+}
+
+func (s *snapshotsAtHeight) shapshots(height uint64) (*g.TransactionStateSnapshot, error) {
+	key := snapshotsKey{height: height}
+	snapshotsBytes, err := s.hs.newestTopEntryData(key.bytes())
+	if err != nil {
+		return nil, err
+	}
+	var res g.TransactionStateSnapshot
+	if err = protobuf.Unmarshal(snapshotsBytes, &res); err != nil {
```
Comment (nickeskov) [inline pkg/state/snapshot_storage.go:38]: And here we can use `vtprotobuf`.

File: pkg/state/snapshot_storage_test.go [hunk 2]
```diff
@@ -0,0 +1,29 @@
+package state //nolint:testpackage //TODO: check this
```
Comment (nickeskov) [inline pkg/state/snapshot_storage_test.go:1]: Needs to be fixed

File: pkg/state/headers_validation_test.go [hunk 3]
```diff
@@ -141,3 +141,20 @@ func TestHeadersValidation(t *testing.T) {
 	err = st.RollbackToHeight(1)
 	assert.NoError(t, err, "failed to rollback state")
 }
+
+func TestSnapshots(t *testing.T) {
+	blocks, err := readBlocksFromTestPath(blocksNumber)
+	if err != nil {
+		t.Fatalf("Can not read blocks from blockchain file: %v\n", err)
+	}
```
Comment (nickeskov) [inline pkg/state/headers_validation_test.go:149]: `require`?

File: pkg/state/appender.go [hunk 4]
```diff
@@ -452,7 +452,97 @@ func (a *txAppender) handleDefaultTransaction(tx proto.Transaction, params *appe
...
+		return nil, nil, false, errors.New("failed to cast interface transaction to ethereum transaction structure")
+	}
+	kind, errResolveTx := a.ethTxKindResolver.ResolveTxKind(ethTx, params.blockRewardDistributionActivated)
+	if errResolveTx != nil {
+		return nil, nil, false, errors.Wrap(errResolveTx, "failed to guess ethereum transaction kind")
+	}
+	ethTx.TxKind = kind
+	var err error
+	switch ethTx.TxKind.(type) {
+	case *proto.EthereumTransferWavesTxKind, *proto.EthereumTransferAssetsErc20TxKind:
```
Comment (nickeskov) [inline pkg/state/appender.go:475]: I think `default` case should be added.

**Review by nickeskov — CHANGES_REQUESTED** [review]

File: pkg/proto/protobuf_converters.go [hunk 5]
```diff
@@ -60,6 +60,217 @@ func SignedTxFromProtobuf(data []byte) (Transaction, error) {
...
+			err := sn.FromProtobuf(scheme, balance)
+			if err != nil {
+				return err
+			}
+			*res = append(*res, sn)
+		}
+	}
+	return nil
+}
+func leaseBalancesFromProto(scheme Scheme, txSnapshotProto *g.TransactionStateSnapshot, res *[]AtomicSnapshot) error {
```
Comment (nickeskov) [inline pkg/proto/protobuf_converters.go:83]: I see that this function appends new snapshot to the `res`.
On my mind the more convenient function signature is:
```go
func appendBalancesFromProto(
	res []AtomicSnapshot,
	scheme Scheme,
	balances []*g.TransactionStateSnapshot_Balance,
) ([]AtomicSnapshot, error) {
	// ...
	return res, nil
}
```
Issue can be applied for all similar functions.

File: pkg/proto/snapshot_types.go [hunk 6]
```diff
@@ -36,6 +73,46 @@ func (s AssetBalanceSnapshot) IsGeneratedByTxDiff() bool {
...
+	var c ProtobufConverter
+	addr, err := c.Address(scheme, p.Address)
+	if err != nil {
+		return err
+	}
+	amount := c.amount(p.Amount)
+	if c.err != nil {
+		return c.err
+	}
+	c.reset()
```
Comment (nickeskov) [inline pkg/proto/snapshot_types.go:96]: Is it necessary? Why do you need `reset` here and for similar cases?

File: pkg/proto/snapshot_types.go [hunk 7]
```diff
@@ -47,6 +124,45 @@ func (s DataEntriesSnapshot) IsGeneratedByTxDiff() bool {
...
+	return nil
+}
+
+func (s *DataEntriesSnapshot) FromProtobuf(scheme Scheme, p *g.TransactionStateSnapshot_AccountData) error {
+	var c ProtobufConverter
+	addr, err := c.Address(scheme, p.Address)
+	if err != nil {
+		return err
+	}
+	var dataEntries []DataEntry
```
Comment (nickeskov) [inline pkg/proto/snapshot_types.go:153]: Maybe preallocate?

File: pkg/proto/snapshot_types.go [hunk 8]
```diff
@@ -59,6 +175,40 @@ func (s AccountScriptSnapshot) IsGeneratedByTxDiff() bool {
...
+func (s AccountScriptSnapshot) AppendToProtobuf(txSnapshots *g.TransactionStateSnapshot) error {
+	snapshotInProto, err := s.ToProtobuf()
+	if err != nil {
+		return err
+	}
+	txSnapshots.AccountScripts = append(txSnapshots.AccountScripts, snapshotInProto)
+	return nil
+}
+
+func (s *AccountScriptSnapshot) FromProtobuf(_ Scheme, p *g.TransactionStateSnapshot_AccountScript) error {
```
Comment (nickeskov) [inline pkg/proto/snapshot_types.go:195]: Remove the unused parameter or initialize `ProtobufConverter` with it.

File: pkg/proto/snapshot_types.go [hunk 9]
```diff
@@ -107,6 +320,71 @@ func (s LeaseStateSnapshot) IsGeneratedByTxDiff() bool {
...
+				Sender:    status.Sender.Body(),
+				Recipient: status.Recipient.Body(),
+			},
+		}
+	case *LeaseStatusCancelled:
+		res.Status = &g.TransactionStateSnapshot_LeaseState_Cancelled_{
+			Cancelled: &g.TransactionStateSnapshot_LeaseState_Cancelled{},
+		}
+	default:
+		return nil, errors.Errorf("Failed to serialize LeaseStateSnapshot to Proto: invalid Lease status")
```
Comment (nickeskov) [inline pkg/proto/snapshot_types.go:342]: Error messages should not begin with a capital letter.

File: pkg/proto/snapshot_types.go [hunk 10]
```diff
@@ -194,6 +663,34 @@ func (s TransactionStatusSnapshot) IsGeneratedByTxDiff() bool {
...
+func (s *TransactionStatusSnapshot) FromProtobuf(_ Scheme, p g.TransactionStatus) error {
+	switch p {
+	case g.TransactionStatus_SUCCEEDED:
+		s.Status = TransactionSucceeded
+	case g.TransactionStatus_FAILED:
+		s.Status = TransactionFailed
+	case g.TransactionStatus_ELIDED:
+		s.Status = TransactionElided
+	default:
+		return errors.Errorf("Undefinded tx status %d", p)
```
Comment (nickeskov) [inline pkg/proto/snapshot_types.go:675]: Capital letter

File: pkg/proto/snapshot_types.go [hunk 11]
```diff
@@ -194,6 +663,34 @@ func (s TransactionStatusSnapshot) IsGeneratedByTxDiff() bool {
...
+func (s TransactionStatusSnapshot) AppendToProtobuf(txSnapshots *g.TransactionStateSnapshot) error {
+	switch s.Status {
+	case TransactionSucceeded:
+		txSnapshots.TransactionStatus = g.TransactionStatus_SUCCEEDED
+	case TransactionElided:
+		txSnapshots.TransactionStatus = g.TransactionStatus_ELIDED
+	case TransactionFailed:
+		txSnapshots.TransactionStatus = g.TransactionStatus_FAILED
+	default:
+		return errors.Errorf("Undefined tx status %d", s.Status)
```
Comment (nickeskov) [inline pkg/proto/snapshot_types.go:689]: And here capital letter

File: pkg/proto/types.go [hunk 12]
```diff
@@ -3062,6 +3062,18 @@ func (s *ScriptInfo) ToProtobuf() *pb.ScriptData {
 	}
 }
 
+func (s *ScriptInfo) ToScriptResponseProtobuf() *pb.ScriptResponse {
+	if s == nil {
+		return &pb.ScriptResponse{}
+	}
+	return &pb.ScriptResponse{
+		ScriptBytes: s.Bytes,
+		ScriptText:  base64.StdEncoding.EncodeToString(s.Bytes),
```
Comment (nickeskov) [inline pkg/proto/types.go:3071]: Are you sure that `ScriptText` field is base64 encoded data?

File: pkg/state/appender.go [hunk 13]
```diff
@@ -698,16 +739,18 @@ func (a *txAppender) appendBlock(params *appendBlockParams) error {
 			stateActionsCounterInBlock:       stateActionsCounterInBlockValidation,
 			currentMinerPK:                   params.block.GeneratorPublicKey,
 		}
-		if err := a.appendTx(tx, appendTxArgs); err != nil {
-			return err
+		txSnapshots, errAppendTx := a.appendTx(tx, appendTxArgs)
+		if errAppendTx != nil {
+			return errAppendTx
 		}
+		blockSnapshots.TxSnapshots = append(blockSnapshots.TxSnapshots, txSnapshots.regular)
```
Comment (nickeskov) [inline pkg/state/appender.go:746]: On my mind It's better to create an `append` function or method for `proto.BlockSnapshot`.

File: pkg/state/snapshot_storage.go [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+) error {
+	key := snapshotsKey{height: blockHeight}
+	blockSnapshotsBytes, err := txSnapshots.MarshallBinary()
+	if err != nil {
+		return err
+	}
+	return s.hs.addNewEntry(snapshots, key.bytes(), blockSnapshotsBytes, blockID)
+}
+
+func (s *snapshotsAtHeight) shapshots(height uint64) (proto.BlockSnapshot, error) {
```
Comment (nickeskov) [inline pkg/state/snapshot_storage.go:29]: Typo, s**n**apshots.

File: pkg/state/snapshot_storage.go [hunk 15]
```diff
@@ -0,0 +1,40 @@
+package state
+
+import (
+	"github.com/wavesplatform/gowaves/pkg/proto"
+)
+
+type snapshotsAtHeight struct {
+	hs     *historyStorage
+	scheme proto.Scheme
```
Comment (nickeskov) [inline pkg/state/snapshot_storage.go:9]: It's uninitialized. Pass `scheme` as an argument for `snapshotsAtHeight.snapshots` method.

**Review by nickeskov — CHANGES_REQUESTED** [review]
`StateVersion` constant in `pkg/state` package **must** be incremented due to incompatibility with the previous version.

File: pkg/proto/protobuf_converters.go (hunk 5 above)
Comment (Anton-Rampage) [reply pkg/proto/protobuf_converters.go:83]: it is need for linter issues (funclen)

File: pkg/proto/types.go (hunk 12 above)
Comment (Anton-Rampage) [reply pkg/proto/types.go:3071]: Similar to ScriptData

File: pkg/proto/types.go [hunk 16]
```diff
@@ -3062,6 +3063,18 @@ func (s *ScriptInfo) ToProtobuf() *pb.ScriptData {
...
+		Complexity:  int64(s.Complexity),
+		PublicKey:   nil,
```
Comment (nickeskov) [inline pkg/proto/types.go:3074]: Why is that `nil`? Is it intended?

Comment (nickeskov) [reply pkg/proto/types.go:3074]: See #1245

File: pkg/proto/types.go (hunk 12 above)
Comment (nickeskov) [reply pkg/proto/types.go:3072]: Okay, because this is a deprecated field

**Review by nickeskov — APPROVED** [review]

File: pkg/proto/protobuf_converters.go (hunk 5 above)
Comment (alexeykiselev) [reply pkg/proto/protobuf_converters.go:83]: Maybe better introduce an `AtomicSnapshots` type and make this function and other functions like this the methods of this type?

**Review by alexeykiselev — APPROVED** [review]

**Pull request merged**